[Blackjack - Step2] Submit Bojana's mission

bojana-petroska opened the pull request:
#### Hi Dani!

With this PR, I've implemented the Blackjack betting system and (inspired by Jason) I've implemented a State Pattern. I tried my best not to break my existing program while following his steps. However, I'm not sure if I have the most idiomatic or correct implementation of it. I would love to hear your opinion on it. My idea was to practice the pattern, and I believe it got a bit clearer to me, but there are still things that I need to improve.

#### Changes made:
- `State Pattern` - using the State interface and its implementations `FirstTurn`, `Hit`, `Blackjack`, `Bust`, `Stay`. The Participant class now delegates hand state management to these state objects.
- Model - `Player` now tracks `bettingAmount` and `finalEarnings`. `Dealer` tracks it's `finalEarnings`. The `Dealer.setResultFor` method calculates and updates these monetary values.
- Controller - `BlackJackGame` handles betting input during player creation and triggers the display of final earnings.
- View - `InputView` prompts for bets, and `OutputView` displays final earnings.

#### My open questions:
- I suspect that all monetary actions shouldn't be spread throughout the classes, but rather centralized. Would it also make sense to have a centralized class or object responsible solely for game related calculations?
- The Dealer's logic in `setResultFor` became lengthy and complex, which again is for me not a clear separation of dealer (dealing) and calculations. Should I rethink about a refactoring here?
- Now I have one centralized controller, but would it make sense for each business logic to have their own controller, for example a `PlayersController`?

Thank you very much for your feedback!

SebasNadu committed:
docs(readme): create the initial feature list

SebasNadu committed:
feat(blackjack): add main application entry point

SebasNadu committed:
test(model): add PlayerTest, to check the initializer and to check if the player has blackJack or busts

SebasNadu committed:
feat(model): add Participant abstract class and Player classes that inherit from Participant

SebasNadu committed:
feat(model): add Card Class with Suit and Rank enums and Card test

implement the Suit Enum with the card's symbol and the Rank Enum the number's symbol and their value

SebasNadu committed:
feat(model): add Deck Class that contains a list of cards and create a DeckTest

it constructs a deck by creating a list of 52 shuffled cards

SebasNadu committed:
feat(model): add Hand class to manage dealt cards and HandTest for validation

implement Hand class to track and evaluate card scores, including methods for blackjack and bust detection. add HandTest to validate card scoring logic and functionality.

SebasNadu committed:
docs(readme): update feature list to reflect recent model implementation progress

SebasNadu committed:
refactor(model): delegate card and score management to Hand class

Refactor `Participant` to use the `Hand` class for card and score management. Adjust related tests to validate `Hand` functionality. Simplified logic in `Participant` for cleaner design.

SebasNadu committed:
feat(model): add a Dealer Class inherited from the abstract Participant Class and add dealCard and shouldStand methods

SebasNadu committed:
feat(input): add getPlayersNames and getAnswer

SebasNadu committed:
feat(output): add printParticipantsHands and getPlayersNames

SebasNadu committed:
feat(controller): add start of the game, createPlayers and dealFirstHands

SebasNadu committed:
feat(model): add override toString logic in Dealer

SebasNadu committed:
refactor(model): change property from private to protected

SebasNadu committed:
docs(readme): update progress on Player and Dealer class features

SebasNadu committed:
refactor(model): rename methods for clarity in Hand and Dealer classes

Renamed `isBlackjack` to `hasBlackJack` in `Hand` and `shouldStand` to `shouldNotStand` in `Dealer` for improved readability. Updated related test cases accordingly.

SebasNadu committed:
refactor(input): add player name to getAnswer prompt for clarity

SebasNadu committed:
feat(output): add methods to print dealer's actions and final hands

Add `printDealersMessage` to display dealer's drawn card message and `printFinalHands` to show final hands with scores of players and dealer.

SebasNadu committed:
feat(controller): implement game logic to determine outcomes

Add methods to handle card dealing for players and dealer. Implement logic to calculate game outcomes, determining wins, losses, and ties based on scores.

SebasNadu committed:
feat(controller): implement game logic to determine outcomes

Add methods to handle card dealing for players and dealer. Implement logic to calculate game outcomes, determining wins, losses, and ties based on scores.

SebasNadu committed:
refactor(controller): extract result calculation logic into separate methods

Moved game outcome determination logic into `calculateResults` and `setResultFrom` methods for improved readability and maintainability. Encapsulated dealer's card visibility with `showAllCards` method.

SebasNadu committed:
docs(readme): update implementation progress for Controller and View components

SebasNadu committed:
feat(model): add getResults method to Participant and overrides in Player, Dealer

Introduce abstract `getResults` method in `Participant`, with custom implementations in `Player` and `Dealer` to format game results. Add `setResultFor` in Dealer to determine outcomes against players.

SebasNadu committed:
refactor(controller): streamline result calculation using Dealer's setResultFor

Simplify `calculateResults` by delegating player outcome determination to Dealer's `setResultFor` method. Remove redundant `setResultFrom` logic and enhance readability.

SebasNadu committed:
feat(output): add printResults method to display final game results

Introduce `printResults` in `OutputView` to showcase the final results for both the dealer and players, enhancing game result presentation.

SebasNadu committed:
feat(view, controller): enhance input validation and dealer message handling

Add non-blank validation for player names in `InputView`. Introduce retry mechanism in `BlackJackGame` for resilient input prompts. Refactor dealer message handling in `OutputView` by splitting into `printDealersDrawMessage` and `printDealersStandMessage` for improved clarity.

SebasNadu committed:
refactor(deck, docs): improve deck exception handling and update README progress

Update `Deck` to provide detailed exception messages when no cards are available. Mark completed tasks in the README, including user input validation with retry logic and game step formatting.

bojana-petroska committed:
init

SebasNadu committed:
docs(readme): update todo section with new tasks for exceptions, prompts, and card colors

bojana-petroska committed:
Merge remote-tracking branch 'sebas/step1' into step1

bojana-petroska committed:
style: format all the files following kotlin convention rules

bojana-petroska committed:
refactor(controller, model): add Players class to encapsulate player related logic

- Moved logic related to managing multiple players from BlackJackGame into a new Players class.
- This improves separation of concerns and prepares the codebase for easier expansion and testing.

bojana-petroska committed:
refactor(model, view): move result formatting from model to OutputView

- Removed getResults() from Player and Dealer model classes to keep models free from view logic
- Added printPlayersResults() and printDealerResults() to OutputView to handle result display
- Updated BlackJackGame to call OutputView for printing results

bojana-petroska committed:
refactor(model, view): move card display logic from model to OutputView for better separation of concerns

- Removed toString override from Participant.
- Updated OutputView to handle dealer and player hand formatting explicitly.

bojana-petroska committed:
refactor(view): add readlnOrNull instead of readLine to prevent a NullPointerException

bojana-petroska committed:
test(model): add unit tests for Dealer class

bojana-petroska committed:
refactor(model, view): Implement State pattern for game logic

Adopted the State design pattern to manage the various states of a participant's hand (FirstTurn, Hit, Blackjack, Bust, Stay). This refactoring improves modularity, maintainability, and extensibility of the game's core logic.

 Key changes include:
 - Introduced `State` interface, `Running`, and `Finished` abstract classes, along with concrete state implementations (`FirstTurn`, `Hit`, `Blackjack`, `Bust`, `Stay`).
 - Refactored `Participant` (and its subclasses `Player`, `Dealer`) to manage `handState` using the new State objects, delegating state transitions and status checks.
 - Simplified `Hand` class to focus solely on card collection and total score calculation, removing game-specific logic (e.g., `isBusts`, `hasBlackJack`).
 - Integrated polymorphic `Card` hierarchy (`abstract Card`, `NumberCard`, `FaceCard`, `AceCard`) with the `Rank` enum for cleaner card creation.
 - Updated `Deck` to instantiate cards using the new polymorphic `Card` types.
 - Streamlined game flow in `Players` and `BlackJackGame` by leveraging the `Participant`'s state management.
 - Corrected a syntax error in `Participant.addCard` (lambda assignment) that prevented state updates.
 - Removed unused secondary constructor in `FirstTurn` for code clarity.

 This change makes the game logic more explicit and easier to reason about, aligning with a robust object-oriented design.

bojana-petroska committed:
refactor(model): Separate Model and View/Controller Concerns

bojana-petroska committed:
docs(readme): update feature list to reflect recent model implementation progress

bojana-petroska committed:
docs(readme): update progress on Player and Dealer class features

bojana-petroska committed:
refactor(input, output): add input for betting and display the final earnings

bojana-petroska committed:
refactor(model): implement betting and earnings tracking

bojana-petroska committed:
refactor(controller): integrate input and final earnings display

bojana-petroska committed:
Merge branch 'bojana-petroska' into step2

da-nyee commented on src/main/kotlin/blackjack/model/Card.kt:
```diff
@@ -1,31 +1,53 @@
+    ACE("A", 1),
+}
+
+abstract class Card(val suit: Suit, val rank: Rank) {
```
<img width="4320" height="2478" alt="image" src="https://github.com/user-attachments/assets/71dda4af-7350-4490-831a-b1856e40c586" />

Could you check `CardTest`?
Since `Card` is an abstract class, it can’t be instantiated 👀

da-nyee commented on src/main/kotlin/blackjack/controller/BlackJackGame.kt:
```diff
@@ -16,18 +16,41 @@ class BlackJackGame {
         players.calculateResults(dealer)
-        OutputView.printResults(players.toList(), dealer)
+        OutputView.printWinnings(players.toList(), dealer)
     }
```
<img width="4320" height="2478" alt="image" src="https://github.com/user-attachments/assets/a998c758-2861-4811-9180-b17f86e9dd01" />

I guess in this case, the `Final Earnings` would be:

```
Dealer: -30000.0  
pobi: 20000.0  
jason: 40000.0
```

<br/>

This is based on the following functional requirement:

```
If the dealer busts (exceeds 21),
all remaining players automatically win and receive payouts based on their bets.
```

da-nyee replied:
@bojana-petroska I think this one hasn’t been fixed yet — could you take another look? 👀

bojana-petroska replied:
<img width="431" height="523" alt="Screenshot 2025-07-19 at 10 15 27" src="https://github.com/user-attachments/assets/12c750d9-df80-40bb-a80f-6be2158d8baf" />


I think it is working properly, or?

da-nyee replied:
Yeah, I think so !

da-nyee commented on src/main/kotlin/blackjack/controller/BlackJackGame.kt:
```diff
@@ -16,18 +16,41 @@ class BlackJackGame {
+        OutputView.printWinnings(players.toList(), dealer)
     }
 
     private fun createPlayers() {
```
Maybe consider delegating this responsibility to `Players`, if that makes sense.

da-nyee commented on src/main/kotlin/blackjack/controller/BlackJackGame.kt:
```diff
@@ -16,18 +16,41 @@ class BlackJackGame {
         players = Players(playerList)
     }
 
+    fun dealingPlayersCards() {
```
This method could be `private`.

da-nyee commented on src/main/kotlin/blackjack/model/Dealer.kt:
```diff
@@ -3,53 +3,69 @@ package blackjack.model
                 this.setLose()
+                player.finalEarnings = playerRawProfit - player.bettingAmount
+                finalEarnings -= playerRawProfit
             }
```
These lines are duplicated with lines 57-62.
Could you refactor the code to remove the duplication?

da-nyee commented on src/main/kotlin/blackjack/model/Dealer.kt:
```diff
@@ -3,53 +3,69 @@ package blackjack.model
-    fun getDealtCards() = hand.dealtCards.toList()
+    fun getDealtCards() = handState.hand.cards.toList()
 
     fun setResultFor(player: Player) {
```
This is a bit too long — could you simplify it or break it down into smaller methods?

According to the programming requirement:
```
Each function or method should not exceed 10 lines.
```

da-nyee commented on src/main/kotlin/blackjack/model/GameState.kt:
```diff
@@ -0,0 +1,70 @@
+package blackjack.model
+
+interface State {
```
I see you applied the State Pattern — well done 👍

da-nyee commented on src/main/kotlin/blackjack/view/InputView.kt:
```diff
@@ -26,4 +26,14 @@ object InputView {
+
+    fun getBettingAmount(playerName: String): Int {
+        println("Enter $playerName's betting amount:")
+        val inputString = readLine()?.trim() ?: ""
```
```suggestion
        val inputString = readlnOrNull()?.trim() ?: ""
```

Consider using this to avoid a potential `NullPointerException`.

da-nyee commented on src/main/kotlin/blackjack/model/GameState.kt:
```diff
@@ -0,0 +1,70 @@
+    }
+}
+
+class FirstTurn(override val hand: Hand = Hand(emptyList())) : Running() {
```
I'd recommend adding unit tests for each concrete state class 👀

da-nyee reviewed (changes requested):
Hi Bojana! 🙌
Great work on this step 👍 

I left some comments that might help improve the code.
Feel free to check them out, and let me know if anything's unclear.

---

> I suspect that all monetary actions shouldn't be spread throughout the classes, but rather centralized. Would it also make sense to have a centralized class or object responsible solely for game related calculations?

Yes!
I think it's a good idea to introduce a new object responsible for handling the calculations.

<br/>

> The Dealer's logic in `setResultFor` became lengthy and complex, which again is for me not a clear separation of dealer (dealing) and calculations. Should I rethink about a refactoring here?

Yeah, I think so.
I'd recommend refactoring the method — I actually already left a similar comment there, though.

<br/>

> Now I have one centralized controller, but would it make sense for each business logic to have their own controller, for example a `PlayersController`?

Well, that makes sense, but I think a single `Controller` is enough for now.
You could consider splitting it into more specific `Controller`s in the next mission when working with Spring / Spring Boot.

bojana-petroska committed:
refactor(model): simplify Card model using data class and static all cards, remove Card subclasses

bojana-petroska committed:
Merge remote-tracking branch 'origin/step2' into step2

bojana-petroska committed:
refactor(model): move state and separate in general state, running and finished. Refined the State Pattern

- new package structure
- simplified Card Model as a data Class instead of an abstract
- implement by lazy for ALL_CARDS for immutability
- redifined Participant abstract class. Now Player and Dealer primarely manage their own Hands State
- added generateParticipants

bojana-petroska committed:
refactor(input, output): change readLine into readlnOrNull to avoid potential NullPointerException, edit the display of the cards

bojana-petroska committed:
refactor(game): edit the game logic

- BlackJackController - delegates core game execution to BlackJackGame and result calculation to GameLogic
- created GameLogic Class responsible for calculating the financial outcomes of the game
- emoved outcome calculation logic from Dealer
- Modifies the `finalEarnings` properties on Player and Dealer

bojana-petroska committed:
style: apply ktlintFormat

da-nyee commented on src/main/kotlin/blackjack/model/Card.kt:
```diff
@@ -1,31 +1,15 @@
 package blackjack.model
 
 data class Card(val suit: Suit, val rank: Rank) {
```
I see you’ve made this class instantiable — nice work 👍
But there’s a compile error in `CardTest`.

<img width="1552" height="987" alt="image" src="https://github.com/user-attachments/assets/f040994c-7761-4047-83e9-f128f74f11de" />

I think it should be something like: `val card = Card(Suit.DIAMONDS, Rank.TEN)`

da-nyee replied:
It might be a good idea to run the tests after refactoring 👀

da-nyee commented on src/main/kotlin/blackjack/game/BlackJackGame.kt:
```diff
@@ -0,0 +1,34 @@
+
+class BlackJackGame(
+    private val dealer: Dealer,
+    private val players: List<Player>,
```
How about wrapping this collection like `Players`?

da-nyee commented on src/main/kotlin/blackjack/model/Dealer.kt:
```diff
@@ -1,57 +1,13 @@
-                player.setWin()
-                this.setLose()
-            }
+    fun playTurn(deck: Deck) {
+        while (handState.isRunning() && handState.hand.calculateSum() <= DEALER_STAND) {
+            draw(deck.drawCard())
         }
+        if (handState.isRunning()) stay()
     }
```
You nailed it 👍

da-nyee commented on src/main/kotlin/blackjack/model/Deck.kt:
```diff
@@ -1,32 +1,16 @@
 
-class Deck() {
-    private val _cards: MutableList<Card> = mutableListOf()
+class Deck(private val cards: MutableList<Card> = Card.ALL_CARDS.shuffled().toMutableList()) {
```
What about introducing a value object like `Cards`?

da-nyee commented on src/test/kotlin/blackjack/model/DealerTest.kt:
```diff
@@ -121,5 +128,4 @@ class DealerTest {
         assertThat(dealer.gameResults.loses).isEqualTo(1)
     }
-
 }
```
<img width="1552" height="987" alt="image" src="https://github.com/user-attachments/assets/2b6962f3-44b7-48c0-ac73-cf7d2fe8a62c" />

Could you check all the tests again?
It looks like there are some compile errors 👀

da-nyee reviewed (changes requested):
Hi Bojana! 🙌
Thanks for applying the feedback 🙇‍♀️

There's still some room for improvement, so I've left a few more comments — please check them out when you have time.
I'm happy to help if you have any questions ☺️

bojana-petroska committed:
refactor(model): introduce Players wrapper for better SRP

bojana-petroska committed:
refactor(tests): align unit tests with the refactored architecture

bojana-petroska committed:
feat(tests): add unit tests for the state

bojana-petroska committed:
refactor(model): change into a data class, remove currentCards

bojana-petroska commented:
Hi @da-nyee !
Thank you very much for the comprehensive feedback that made me revisit my code many times and try to improve it. I also took some more time to work on it, in order to understand the State better. I hope I have some better structure now. I tried to follow the requirements and kotlin conventions (ok maybe not fully TDD :S). Before requesting the review, I checked all the tests and the app are running : )

da-nyee commented on src/main/kotlin/blackjack/game/BlackJackGame.kt:
```diff
@@ -0,0 +1,31 @@
+class BlackJackGame(
+    private val dealer: Dealer,
+    private val players: Players,
+    private val getAnswer: (Player) -> Boolean,
+    private val displayPlayerHand: (Player) -> Unit,
```
You used lambda expressions 👍

da-nyee commented on src/main/kotlin/blackjack/game/GameLogic.kt:
```diff
@@ -0,0 +1,28 @@
+            val playerFinalState = player.handState as Finished
+
+            val playerProfit = playerFinalState.profit(dealerFinalState, player.bettingAmount)
+            player.finalEarnings = playerProfit
```
You might consider sending a message to the player instead of setting the data directly here 👀

da-nyee commented on src/main/kotlin/blackjack/state/State.kt:
```diff
@@ -0,0 +1,24 @@
+import blackjack.model.Card
+import blackjack.model.Hand
+
+interface State {
```
You implemented it very well 👍

da-nyee commented on src/test/kotlin/state/BlackjackTest.kt:
```diff
@@ -0,0 +1,83 @@
+import org.junit.jupiter.api.assertThrows
+import kotlin.test.Test
+
+class BlackjackTest {
```
I see you've tested the features thoroughly — great work 👍

da-nyee reviewed (approved):
Hi Bojana! 🙌
I've gone through all the changes — thanks for applying the feedback 🙇‍♀️

I can really see your growth — your coding skills are getting better and better!
It's time to wrap up the mission. You made it 💯

I hope the feedback was helpful, and I also hope you're enjoying the process of learning and coding.
If you have any questions, feel free to reach out — I'm always happy to help ☺️
